ironfish/src/blockchain/nullifierSet/nullifierSet.ts

```diff
@@ -0,0 +1,132 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+import { Assert } from '../../assert'
+import { Block, Transaction } from '../../primitives'
+import { Nullifier } from '../../primitives/nullifier'
+import {
+  BUFFER_ENCODING,
+  DatabaseSchema,
+  IDatabase,
+  IDatabaseStore,
+  IDatabaseTransaction,
+  StringEncoding,
+  U32_ENCODING,
+} from '../../storage'
+import { NullifierInfo, NullifierLocationEncoding } from './encoding'
+
+interface CounterEntry<T extends string> extends DatabaseSchema {
+  key: T
+  value: number
+}
+
+interface NullifiersSchema extends DatabaseSchema {
+  key: Nullifier
+  value: NullifierInfo
+}
+
+export class NullifierSet {
+  readonly db: IDatabase
+
+  // Keep track of number of nullifiers in the set
+  private readonly counter: IDatabaseStore<CounterEntry<'Size'>>
+
+  // Nullifier -> BlockHash, TransactionHash
+  private readonly nullifiers: IDatabaseStore<NullifiersSchema>
+
+  constructor(options: { db: IDatabase; storeName: string }) {
+    this.db = options.db
+
+    this.counter = this.db.addStore({
+      name: `${options.storeName}c`,
+      keyEncoding: new StringEncoding<'Size'>(),
+      valueEncoding: U32_ENCODING,
+    })
+
+    // Nullifier -> BlockHash, TransactionHash
+    this.nullifiers = this.db.addStore({
+      name: options.storeName,
+      keyEncoding: BUFFER_ENCODING,
+      valueEncoding: new NullifierLocationEncoding(),
+    })
+  }
+
+  size(tx?: IDatabaseTransaction): Promise<number> {
+    return this.db.withTransaction(tx, async (tx) => {
```

Comment:
why is this starting a transaction?

Reply:
So we can call contains without passing in a transaction if we need to

Reply:
As you pointed out, `db.get` can also take an undefined transaction and will create its own so removing this.